src/core/search/search.cc

```diff
@@ -622,6 +620,44 @@ void FieldIndices::FinalizeInitialization() {
   }
 }
 
+namespace {
+
+template <typename T>
+DefragmentResult DefragmentIndices(T& indices, PageUsage* page_usage, size_t quota,
```

Comment:
This ended up being very similar to `ShardDocIndices::Defragment`, maybe they can be refactored into one function. RaxTree defragmentation is also similar but slightly different as that iterator cannot be moved or copied.

Reply:
all defrag-map style code is refactored into one class now